src/bare-metal/microcontrollers/examples/src/bin/mmio.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ANCHOR: Example
+#![no_main]
+#![no_std]
+
+extern crate panic_halt as _;
+
+use core::mem::size_of;
+use cortex_m_rt::entry;
+
+/// GPIO port 0 peripheral address
+const GPIO_P0: usize = 0x5000_0000;
+
+// GPIO peripheral offsets
+const PIN_CNF: usize = 0x700;
+const OUTSET: usize = 0x508;
+const OUTCLR: usize = 0x50c;
+
+// PIN_CNF fields
+const DIR_OUTPUT: u32 = 0x1;
+const INPUT_DISCONNECT: u32 = 0x1 << 1;
+const PULL_DISABLED: u32 = 0x0 << 2;
+const DRIVE_S0S1: u32 = 0x0 << 8;
+const SENSE_DISABLED: u32 = 0x0 << 16;
+
+#[entry]
+fn main() -> ! {
+    // Configure GPIO 0 pins 21 and 28 as push-pull outputs.
+    let pin_cnf_21 = (GPIO_P0 + PIN_CNF + 21 * size_of::<u32>()) as *mut u32;
+    let pin_cnf_28 = (GPIO_P0 + PIN_CNF + 28 * size_of::<u32>()) as *mut u32;
+    // Safe because the pointers are to valid peripheral control registers, and no aliases exist.
+    unsafe {
+        pin_cnf_21.write_volatile(
+            DIR_OUTPUT | INPUT_DISCONNECT | PULL_DISABLED | DRIVE_S0S1 | SENSE_DISABLED,
+        );
+        pin_cnf_28.write_volatile(
+            DIR_OUTPUT | INPUT_DISCONNECT | PULL_DISABLED | DRIVE_S0S1 | SENSE_DISABLED,
+        );
+    }
+
+    // Set pin 28 high and pin 21 low to turn the LED on.
+    let gpio0_outset = (GPIO_P0 + OUTSET) as *mut u32;
+    let gpio0_outclr = (GPIO_P0 + OUTCLR) as *mut u32;
+    // Safe because the pointers are to valid peripheral control registers, and no aliases exist.
```

Comment:
The line is clipped here:

![image](https://user-images.githubusercontent.com/89623/222719903-002c1118-b44b-4be3-be3c-f1b6e2327698.png)

Reply:
Wrapped at 80 columns.